src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -48,14 +53,12 @@ internal partial class DiagnosticIncrementalAnalyzer
                     return existingData;
                 }
 
-                // Perf optimization: Check whether analyzer is suppressed for project or document and avoid getting diagnostics if suppressed.
-                if (!DocumentAnalysisExecutor.IsAnalyzerEnabledForProject(stateSet.Analyzer, document.Project, GlobalOptions) ||
+                // Check whether analyzer is suppressed for project or document and avoid getting diagnostics if suppressed.
+                // Note that we do not return empty DocumentAnalysisData for suppressed analyzer as that would clear the
+                // error list and remove the previously reported diagnostics from "Run Code Analysis" command.
```

Comment:
i think this is a bit too vague here.  I have no idea in the context of this code what the "run code analysis" command is or why it has this special capability.  What *does* end up clearing the results of it?  something feels *very* stateful here wrt to the error list and thsee features, adn i don't have a clear sense of how that state is managed and what is or isn't ok.

Note: this also seems like a scenario that i have no idea how it will work with pull diags.

Reply:
> I have no idea in the context of this code what the "run code analysis" command is or why it has this special capability. What does end up clearing the results of it?

I'll end up adding a better comment here. "Run code analysis" is just one of the scenarios, but the core issue was prior code was essentially stating - there are no diagnostics in the document that we are trying to fetch cached diagnostics for, which was not true. This ends up pushing empty diagnostics for that document to the error list. The new code is basically returning null (indicating there are no cached diagnostics) and setting a flag that clearly indicates that analysis is disabled for this document, so the caller is free to perform any optimizations around analysis of that document, such as skipping running analysis for it.

Reply:
> Note: this also seems like a scenario that i have no idea how it will work with pull diags.

Tracked by https://github.com/dotnet/roslyn/issues/65967. However, given the large number of customer reports about this regression, we ought to fix this asap for push diagnostics, we cannot wait until pull diagnostics are enabled and #65967 is fixed.

Reply:
> i think this is a bit too vague here. I have no idea in the context of this code what the "run code analysis" command is or why it has this special capability. What _does_ end up clearing the results of it? something feels _very_ stateful here wrt to the error list and thsee features, adn i don't have a clear sense of how that state is managed and what is or isn't ok.
> 
> Note: this also seems like a scenario that i have no idea how it will work with pull diags.

As a small-scale self-employed nobody who actually uses your VS Community tool and doesn't work for Microsoft, or have friends at MS, but *does* use VS for 80% of the programming work that I do in my career -- let me explain the issue. Listen to your user. (You used to call us "Elvises", remember that?)

My assumption (and forgive me if it's wrong, but explain why --) is that you added this code analysis feature for Enterprise users, and we get it for free in Community (which is amazing, I love VS) but it wasn't properly tested there.

Here is how I can *repeatably* reproduce the annoyance. I open my recent VS (it is 2022 64-bit, 17.4.3, with .NET Framework 4.8.04084). I load my primary work project, it's nothing magic, it's just a nice large Windows Forms application that I began in 2016. And I do: Analyze (menu) > Code Analysis > On Solution. Then I am shown the useful output. But once I double-click any entry and start to look at a single source file, then all the other analysis (other files) disappears.

Thanks for listening to your user.

Reply:
@equin0x80 my messages were about lack of clarity on what this fix was doing. Thanks!

Reply:
I have clarified the comment and also have a descriptive doc comment on the method.